### PR TITLE
[membership] Add a membership based voting strategy

### PR DESCRIPTION
**Pull request by blakewest**
Changes proposed in this pull request:
- Add a new strategy called `membership`
- This strategy will let you use any "membership strategy" along with any "voting power strategy". The membership strategy weeds out addresses to only those that pass the membership strategy (ie. return a number > 0), and then passes those on to the voting power strategy. 
- Both strategies must be already existing snapshot strategies.
- This is useful for using an identity NFT (such as [UID](https://etherscan.io/address/0xba0439088dc1e75F58e0A7C107627942C15cbb41)) in order to enable things like quadratic voting with on-chain sybil protection. But the design space is very wide, and many other forms of membership could be used, such as on-chain behavior from a subgraph, or owning a PUNK, or BrightID verification.

### CODE REVIEW

**Comment by blakewest** [comment]
@ChaituVR I see the `changes-requested` tag, but I don't see any comments or changes that have been requested. Can you clarify what you'd like to see?

**Review by ChaituVR — CHANGES_REQUESTED** [review]

File: src/strategies/erc20-balance-of-nft-gated/index.ts [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        {
+          internalType: 'uint256[]',
+          name: '',
+          type: 'uint256[]'
+        }
+      ],
+      stateMutability: 'view',
+      type: 'function'
+    }
+  ];
```
[the remark this answers is not in the document]
Comment (ChaituVR) [reply src/strategies/erc20-balance-of-nft-gated/index.ts:95]: @blakewest sorry :) forgot to publish the review (small change)

**Comment by blakewest** [comment]
@ChaituVR updated!

**Review by bonustrack — COMMENTED** [review]
It would be great if you make this strategy more general purpose, currently it seem to work only for ERC20 as voting power with NFT validation, but ideally it should be made in a way that you could use any strategy as voting power and any strategy as validation. For example we could use ERC721 strategy as voting power and BrightID strategy as validation. I would recommend use a structure like this:
```
{
   votingPowerStrategy: {
      name: 'erc20-balance-of',
      params: {
         address: '0x123243...',
         decimals: 18
      }
   }
   validationStrategy: {
      name: 'brightid',
      params: {}
   }
}
```
 This would give much more freedom

**Comment by blakewest** [comment]
@bonustrack ah, good call! I dig that. I'll take a crack at it.

**Comment by blakewest** [comment]
@bonustrack and @ChaituVR updated!

**Review by bonustrack — COMMENTED** [review]
@blakewest Looks good, could you just rename the strategy to just "membership"?

**Comment by blakewest** [comment]
@bonustrack I updated the name. Though I worry the name is not very descriptive, and people might confuse it with something where your voting power is like 1 or 0 based on a membership. As I write this out, do you think `members-only-voting` would be better? I'm also fine to just get this merged in. We'd like to use this very soon, so I want to make sure it's in, and we can do some initial tests.

**Comment by blakewest** [comment]
@ChaituVR I see a couple tests are failing. But looks like it's from a server error on the specs. Can you re-run? Or are we good to merge as is? Let me know if I can help. Thanks! - Blake

**Comment by bonustrack** [comment]
@blakewest Hmm I don't see so much differences from membership-based-voting or members-only-voting to membership name. Each of them could be confused with 0-1 type of strategy. Ideally I'd like to keep the name as generic as possible so it's easy to remember, I believe this can be a very popular strategy.

**Comment by blakewest** [comment]
Alright. Sounds good, let's get it merged!

**Pull request merged**